### ISSUE TITLE
Is `waiterList.Remove(elem)` in `Acquire` always work correctly?

### ISSUE THREAD

**Issue by blastbao** [sherifabdlnaby/semaphore#1]
hi,sorry for bothering, I have a question, If one calls Acquire with N bigger than s.size, then it will be appended to `s.impossibleWaiters` list, but after a while Resize could move it from `s.impossibleWaiters` to `s.waiters`, then when the caller's ctx is canceled, the `waiterList.Remove(elem)` cannot remove it correctly, may this cause resource leak?

**Comment by sherifabdlnaby** [comment]
@blastbao Yes, that is a nice catch, it should remove it from both lists. I will try to fix it soon. 
I am curious that this fork is being explored, I almost forgot about this PR, care to mention why do you need it? :D
